### ISSUE TITLE
Must not cache interrupted download

### ISSUE THREAD

**Issue by mkarg** [download-maven-plugin/download-maven-plugin#208]
When killing the Maven process while download is in progress, the broken file is kept in cache. When restarting the build, the broken file is used just as if it would be complete. To prevent this, incomplete downloads should never be found in cache.

**Comment by mkarg** [comment]
@longtimeago Any trick how I can clean the download cache, so I can remove the broken file from there...?

**Comment by longtimeago** [comment]
@mkarg manually you can just drop folder `~/.m2/repository/.cache/download-maven-plugin/`

**Comment by longtimeago** [comment]
> When restarting the build, the broken file is used just as if it would be complete

@mkarg to avoid that you can provide expected file signature (hash) and add config param `alwaysVerifyChecksum`.
This way, hash is validated on initial download as well as on subsequent extractions from the cache

**Comment by ajarmoniuk** [comment]
This has been around for ages and version 1.3.0, which we used, actually broke builds: the `0-length` index.ser files would not be deserialized properly and the plugin would barf with the EOFException (thrown by ObjectInputStream).

**Comment by mkarg** [comment]
I am not sure if I got you right: So you want to keep the bug?

**Comment by ajarmoniuk** [comment]
I think you misunderstood me. What I meant to say: it would terminate exceptionally in version 1.3.0 if it saw a 0-byte index.ser file in its cache dir, left after a previous unsuccessful run.

**Comment by ajarmoniuk** [comment]
@longtimeago hope you're safe. Let me know if I could fix this.

**Comment by ajarmoniuk** [comment]
Maybe a small update now that the change has been merged.

It simply stores its cache entries in a directory in the local repository (which @mkarg  probably refers to as "Maven cache") folder by default. I've set off with an experiment of migrating to Apache HttpClient cache to see if it's feasible and how much could be gained from it.

Well, Apache HttpClient does provide some means of managing cache entries like invalidation etc, but it does not come with any facilities to load and update the index of the file-based cache. It's up to the user. Also, Apache HttpClient might be a bit to strict wrt to the HTTP spec as it will only cache entries which are explicitly allowed to be cached... which is not always the cache with large files. 

The biggest problem with Apache HttpClient Cache is that I think it's intended use case is completely different than here. What we want here is a one time load off the file storage and update the cache whereas the library is more geared towards multiple usage during the cache instance lifetime. With one-off usage it's quite similar to what is currently implemented here.
